lib/pbench/server/api/resources/users_api.py

```diff
@@ -0,0 +1,534 @@
+        except EmailNotValidError:
+            self.logger.warning("Invalid email {}", email_id)
+            abort(400, message=f"Invalid email: {email_id}")
+        except Exception:
+            self.logger.exception("Exception while registering a user")
+            abort(500, message="INTERNAL ERROR")
+
+
+class Login(Resource):
+    """
+    Pbench API for User Login or generating an auth token
+    """
+
+    def __init__(self, config, logger, auth):
+        self.server_config = config
+        self.logger = logger
+        self.auth = auth
+        try:
+            self.token_expire_duration = self.server_config.get(
+                "pbench-server", "token_expiration_duration"
+            )
+        except NoOptionError:
+            self.token_expire_duration = None
+
+    @Auth.token_auth.login_required(optional=True, f=Auth().verify_auth())
+    def post(self):
+        """
+        Post request for logging in user.
+        The user is allowed to re-login multiple times and each time a new valid auth token will be provided
+
+        This requires a JSON data with required user metadata fields
+        {
+            "username": "username",
+            "password": "password",
+        }
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+
+        :return: JSON Payload
+        if we succeed to decrypt the password hash, the returned response_object will include the auth_token
+            response_object = {
+                    "status": "success", # will not present if failed
+                    "message": "Successfully logged in."/"failure message",
+                    "auth_token": "<authorization_token>", # Will not present if failed
+                }
+        """
+        # get the post data
+        post_data = request.get_json()
+        if not post_data:
+            self.logger.warning("Invalid json object: {}", request.url)
+            abort(400, message="Invalid json object in request")
+
+        username = post_data.get("username")
+        if not username:
+            self.logger.warning("Username not provided during the login process")
+            abort(400, message="Please provide a valid username")
+
+        password = post_data.get("password")
+        if not password:
+            self.logger.warning("Password not provided during the login process")
+            abort(400, message="Please provide a valid password")
+
+        try:
+            # fetch the user data
+            user = User.query(username=username)
+        except Exception:
+            self.logger.exception("Exception occurred during user login")
+            abort(500, message="INTERNAL ERROR")
+
+        if not user:
+            self.logger.warning(
+                "No user found in the db for Username: {} while login", username
+            )
+            abort(403, message="No such user, please register first")
```

Comment:
Or we could just say what happened without explicitly telling them what to do like just  `abort(403, message="No such user")`?

Reply:
I think when it comes to security, "coy" is best. For example, no clue whether they guessed username or password or both incorrectly. Just "Do not pass Go, do not collect $200".

Reply:
The problem is, if we report "no such user", then that potentially enables a brute force attack on the user list; then the malactor, once it locates a valid username, moves on to a brute force attack on the password.  If we report the same error for bad username and bad password, it vastly increases the search space, thereby defeating the naive brute force attack.

So, yes, in this instance, we _should_ be coy.  (Once the user is authenticated, however, I think we should try to be detailed and helpful with our error messages, because, at that point, the user is our friend...or even us!)

Reply:
Fixed